proposals/mutatingpolicies.md

```diff
@@ -0,0 +1,159 @@
+                Object.spec.initContainers{
+                  name: "mesh-proxy",
+                  image: "mesh/proxy:v1.0.0",
+                  args: ["proxy", "sidecar"],
+                  restartPolicy: "Always"
+                }
+              ]
+            }
+          }          
+```
+
+```yaml
+apiVersion: policies.kyverno.io/v1alpha1
+kind: MutatingPolicy
+metadata:
+  name: "sidecar-policy.example.com"
+spec:
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   [""]
+      apiVersions: ["v1"]
+      operations:  ["CREATE"]
+      resources:   ["pods"]
+  matchConditions:
+    - name: does-not-already-have-sidecar
+      expression: "!object.spec.initContainers.exists(ic, ic.name == \"mesh-proxy\")"
+  failurePolicy: Fail
+  reinvocationPolicy: IfNeeded
+  mutations:
+    - patchType: "JSONPatch"
+      jsonPatch:
+        expression: >
+          [
+            JSONPatch{
+              op: "add", path: "/spec/initContainers/-",
+              value: Object.spec.initContainers{
+                name: "mesh-proxy",
+                image: "mesh-proxy/v1.0.0",
+                restartPolicy: "Always"
+              }
+            }
+          ]          
+```
+
+### Mutate Existing
+
+In order to mutate existing resources, a `targets` object is added to custom existing resources:
+
+* `targets.matchConstraints`: required, selects existing resources to mutate based on GVR, `namespaceSelector`, `objectSelector`, etc.
+* `targets.matchConditions`: optional, filters target via CEL-based conditions.
+
+By default `spec.mutateExistingOnPolicyUpdate` is disabled, set it to `true` to mutate existing resources based on policy events.
+
+```yaml
+apiVersion: policies.kyverno.io/v1alpha1
+kind: MutatingPolicy
+metadata:
+  name: add-label-to-existing-secret
+spec:
+  mutateExistingOnPolicyUpdate: true
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   [""]
+      apiVersions: [v1]
+      operations:  [CREATE]
+      resources:   ["configmaps"]
+  ## custom targets filters
+  targetMatchConstraints:
+  - # namespaceSelector:
+    # objectSelector:
+    # excludeResourceRules:
+    resourceRules:
+    - apiGroups:   [""]
+      apiVersions: [v1]
+      resources:   ["secrets"]
+  mutations:
+    - patchType: JSONPatch
+      jsonPatch: 
+        expression: >-
+          [
+            JSONPatch{
+              op: "add", path: "/metadata/labels",
+              value: {foo: bar}
+            }
+          ] 
+```
+
+The built-in variable `target` is supported to reference the target object. For example, the variable `target` is used below to filter secret's namespace via `targets.matchConditions`, and to reference `.spec` object of the secret.
+
+```yaml
+  targetsMatchConstraints:
+  - resourceRules:
+    - apiGroups:   [""]
+      apiVersions: [v1]
+      resources:   ["secrets"]
+  targetMatchConditions:
+  - name: "skip kube-system namespace"
+    expression: >-
+      target.metadata.namespace != "kube-system"
```

Comment:
This will return a list of secrets, meaning we'll have multiple targets. Are we planning to loop through each target and apply the mutation individually?

Reply:
Yes, mutations are applied to each target resource individually.

We can have another `mutations.foreach` field to handle loop iterations.